### PR TITLE
Magazines now lazyload their contents in order to save on gun init times

### PR DESCRIPTION
**Pull request by SmArtKar**
## About The Pull Request
Ammo boxes, this includes magazines and magazines inside of guns, now lazyload their ammo in order to save on init times by keeping it as types in their ammo list and initializing them as needed. As a side effect, you can only use get_round to access rounds now, direct array access will probably not work.

## Why It's Good For The Game

trying to save on init times and delay death of guncode by a tiny bit

## Changelog
:cl:
code: Magazines now lazyload their contents in order to save on gun init times. Please report any broken/non-functional guns!
/:cl:

### CODE REVIEW

**Comment by LemonInTheDark** [comment]
https://github.com/tgstation/dev-cycles-initiative/issues/4
also https://github.com/tgstation/dev-cycles-initiative/issues/19

**Comment by necromanceranne** [comment]
Nice nice

File: code/modules/projectiles/boxes_magazines/_box_magazine.dm [hunk 1]
```diff
@@ -104,59 +107,68 @@
...
+		stored_ammo -= casing
+		casing.forceMove(get_turf(src))
 
-				stored_ammo += R
-				R.forceMove(src)
-				return TRUE
-	return FALSE
+		stored_ammo += new_round
+		new_round.forceMove(src)
+		return TRUE
```
Comment (LemonInTheDark) [inline code/modules/projectiles/boxes_magazines/_box_magazine.dm:143]: needs the default case of return FALSE right?

File: code/modules/projectiles/boxes_magazines/_box_magazine.dm [hunk 2]
```diff
@@ -49,7 +49,10 @@
 	update_icon_state()
 
 /obj/item/ammo_box/Destroy(force)
-	QDEL_LIST(stored_ammo)
+	for (var/casing in stored_ammo)
```
Comment (LemonInTheDark) [inline code/modules/projectiles/boxes_magazines/_box_magazine.dm:52]: why type other iterations but not this?

File: code/modules/projectiles/boxes_magazines/_box_magazine.dm [hunk 3]
```diff
@@ -225,18 +237,24 @@
...
-	for(var/obj/item/ammo_casing/bullet in stored_ammo)
-		if(bullet && (bullet.loaded_projectile || countempties))
+	for(var/obj/item/ammo_casing/bullet as anything in stored_ammo)
+		if(ispath(bullet) || bullet && (bullet.loaded_projectile || countempties))
 			boolets++
 	return boolets
 
 ///list of every bullet in the magazine
+///forces all bullets to lazyload
 /obj/item/ammo_box/magazine/proc/ammo_list()
```
Comment (LemonInTheDark) [inline code/modules/projectiles/boxes_magazines/_box_magazine.dm:247]: what calls this?

File: code/modules/projectiles/boxes_magazines/internal/_cylinder.dm [hunk 4]
```diff
@@ -37,14 +42,17 @@
...
-		if(!bullet || !bullet.loaded_projectile) // found a spent ammo
-			stored_ammo[i] = R
-			R.forceMove(src)
-
-			if(bullet)
-				bullet.forceMove(drop_location())
-			return TRUE
+		if (ispath(bullet))
+			continue
+		if(istype(bullet) && bullet.loaded_projectile)
```
Comment (LemonInTheDark) [inline code/modules/projectiles/boxes_magazines/internal/_cylinder.dm:47]: why not !istype in the first conditional here? two birds one stone

File: code/modules/projectiles/guns/ballistic/revolver.dm [hunk 5]
```diff
@@ -26,10 +26,7 @@
 		CRASH("revolver tried to chamber a round without a magazine!")
 	if(chambered)
 		UnregisterSignal(chambered, COMSIG_MOVABLE_MOVED)
-	if(spin_cylinder)
-		chambered = magazine.get_round(TRUE)
-	else
-		chambered = magazine.stored_ammo[1]
+	chambered = magazine.get_round(!spin_cylinder)
```
Comment (LemonInTheDark) [inline code/modules/projectiles/guns/ballistic/revolver.dm:29]: be explicit about args here

File: code/modules/projectiles/boxes_magazines/internal/_cylinder.dm [hunk 6]
```diff
@@ -13,9 +13,14 @@
 			update_appearance()
 			return
 
-/obj/item/ammo_box/magazine/internal/cylinder/get_round()
-	rotate()
-	return stored_ammo[1]
+/obj/item/ammo_box/magazine/internal/cylinder/get_round(keep = FALSE)
```
Comment (LemonInTheDark) [inline code/modules/projectiles/boxes_magazines/internal/_cylinder.dm:16]: this seems a bit clunky, ideally we'd have two procs, one that hands back the top round and one hands back an unowned top round

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

**Comment by LemonInTheDark** [comment]
Does this actually save time? profiled?

**Comment by necromanceranne** [comment]
> Does this actually save time? profiled?

I swear didn't mothblocks already do that? I distinctly remember her talking about this as something affecting init times because of the guns at centcom

File: code/modules/projectiles/boxes_magazines/_box_magazine.dm [hunk 7]
```diff
@@ -95,68 +98,84 @@
...
-	if (keep)
-		stored_ammo -= casing
-		stored_ammo.Insert(1,casing)
+	if (ispath(casing))
+		casing = new casing(src)
+		stored_ammo[ammo_len] = casing
+	return casing
+
+/// Gets a round from the magazine and puts it back at the bottom of the ammo list
+/obj/item/ammo_box/proc/fetch_round()
```
Comment (LemonInTheDark) [inline code/modules/projectiles/boxes_magazines/_box_magazine.dm:116]: get_and_rotate? name is non ideal

Comment (LemonInTheDark) [reply code/modules/projectiles/boxes_magazines/_box_magazine.dm:116]: cannot come up wth a better one tho

**Comment by LemonInTheDark** [comment]
we need to work out wtf is going on with revolvers here and why the proc names have to suck

**Pull request merged**